### ISSUE TITLE
Consider using a truncated SHA-256 / SHAKE hash instead of MD5

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/butler#68]
Although I'm really curious to find a collision between equal-sized blocks of size <=64K for both the weak hash (original rsync rolling hash) *and* MD5, [Henry Goffin](https://twitter.com/BadMetaphor/status/756026082906648576) raised a good point: [truncating the output of a stronger hash](http://crypto.stackexchange.com/questions/25753/has-a-collision-ever-been-found-for-sha-1-2-3-when-truncated-to-128-bits?rq=1) might be a good option.

Back when I designed wharf, I benchmarked SHA-256 and a few SHAKE variants, but they significantly affected throughput, even if computing the hash itself wasn't much slower. Output size means larger signature files, but more importantly, when doing [checksum searching](https://rsync.samba.org/tech_report/node4.html), comparing strong hashes is slower too (and uses more memory, and loses some access locality, etc).

However, maybe using a truncated SHA-2 or a SHAKE hash of 128-bits would be "safer" (fewer known collisions, better theoretical collision resistance) than relying on MD5.

butler's file formats are versioned and extensible, since they're just protobuf messages — it would be feasible for butler to support both hash formats, and the transition would just seamlessly happen:

  * pre-transition: butler downloads md5 sig, computes md5 sig, diffs + uploads md5 sig
  * during transition: butler downloads md5 sig, computes both md5 and new-strong sigs, diffs using md5 and uploads new-strong sig
  * post-transition: butler downloads new-strong sig, computes new-strong sig, diffs + uploads new-strong sig.

The only "breaking" scenario is someone uploading using a newer butler, then trying to push a new build to the same channel using an older butler. However, we can make the API return an error explaining how to upgrade butler (if they're using butler at all — which we know from the user agent) when the new-strong sig is there but md5 isn't.

The backend already supports multiple signature variants, even though only one is used right now.

**Comment by fasterthanlime** [comment]
Actually, no need to truncate with Keccak sponges, since they have arbitrary output lengths.

Picking `x=64` and `y=0 (don't care)` on the [Keccak tune page](http://keccak.noekeon.org/tune.html) gives the following results:

---

The optimal choice of parameters is:

**Keccak[_r_=1472,_c_=128]** with a least **128 bits** of output.

#### Security claim

In line with our [hermetic sponge strategy](http://sponge.noekeon.org/index.html#hermetic), we make a [flat sponge claim](http://sponge.noekeon.org/index.html#flat) with _c_=128 bits of capacity: for any output length, we claim this Keccak sponge function resists any attack up to **2^64** operations (each of complexity equivalent to one call to Keccak-_f_), unless easier on a random oracle. For 128 bits of output specifically, this translates into the following claimed security level:

Claimed security level

Collision resistance 2^64

(Second) preimage resistance 2^64

#### Speed

For long messages, this function is **43.8% faster than **Keccak[] (Keccak with the default parameters). On the reference processor proposed by NIST, long messages should take about **8.8 cycles/byte**.

#### Addendum: how big is 2^64?

If an attacker has access to one billion computers, each performing one billion evaluations of Keccak-_f_ per second, it would take about **18.4 seconds** to evaluate the permutation 2^64 times.

---

For comparison, here's the cycles per byte for a few other hashes, including MD5:

| Algorithm | MiB/second | Cycles Per Byte |
|-----------|------------|-----------------|
| CRC32     | 253        | 6.9             |
| Adler32   | 920        | 1.9             |
| MD5       | 255        | 6.8             |
| SHA-1     | 153        | 11.4            |
| SHA-256   | 111        | 15.8            |
| SHA-512   | 99         | 17.7            |

Note that the [sha3 package in x/crypto](https://godoc.org/golang.org/x/crypto/sha3) is probably a lot slower than [this cgo binding](https://github.com/steakknife/keccak).

**Comment by fasterthanlime** [comment]
Looks like <https://github.com/steakknife/keccak> uses different parameters for SHAKE-224 than the [official Keccak ones](http://keccak.noekeon.org/NoteOnKeccakParametersAndUsage.pdf) — I'd be wary of how up-to-date the repo is, we'll definitely need to compare with the official implementation.

**Comment by fasterthanlime** [comment]
[Added a few benchmarks](https://github.com/fasterthanlime/keccak), run with `go test -v -bench .`

Results on `go1.6.2 darwin/amd64` (mid-2014 MBP):

```
BenchmarkMD5Write1MiB-8             1000           1669941 ns/op         627.91 MB/s
BenchmarkMD5Sum-8                2000000               956 ns/op        1095916.44 MB/s
BenchmarkKeccak128Write1MiB-8        500           2909285 ns/op         360.42 MB/s
BenchmarkKeccak128Sum-8          1000000              2357 ns/op        444747.22 MB/s
BenchmarkShake128128Write1MiB-8      500           3731020 ns/op         281.04 MB/s
BenchmarkShake128128Sum-8        1000000              1482 ns/op        707372.27 MB/s
BenchmarkKeccak224Write1MiB-8        500           3657307 ns/op         286.71 MB/s
BenchmarkKeccak224Sum-8           500000              2316 ns/op        452744.85 MB/s
BenchmarkKeccak256Write1MiB-8        500           3580572 ns/op         292.85 MB/s
BenchmarkKeccak256Sum-8          1000000              2320 ns/op        451896.20 MB/s
BenchmarkKeccak384Write1MiB-8        300           5016867 ns/op         209.01 MB/s
BenchmarkKeccak384Sum-8          1000000              2306 ns/op        454537.74 MB/s
BenchmarkKeccak512Write1MiB-8        200           7157001 ns/op         146.51 MB/s
BenchmarkKeccak512Sum-8           500000              2343 ns/op        447446.42 MB/s
```

**Comment by fasterthanlime** [comment]
Results on `go1.6.2 windows/amd64` (Core i3 540 @ 3.07GHz):

```
BenchmarkMD5Write1MiB-4             1000           1622358 ns/op         646.33 MB/s
BenchmarkMD5Sum-4                2000000               695 ns/op        1507416.57 MB/s
BenchmarkKeccak128Write1MiB-4        500           3789952 ns/op         276.67 MB/s
BenchmarkKeccak128Sum-4           500000              3334 ns/op        314508.90 MB/s
BenchmarkShake128128Write1MiB-4      300           4470407 ns/op         234.56 MB/s
BenchmarkShake128128Sum-4        1000000              1313 ns/op        798519.96 MB/s
BenchmarkKeccak224Write1MiB-4        300           4632288 ns/op         226.36 MB/s
BenchmarkKeccak224Sum-4           500000              2985 ns/op        351232.28 MB/s
BenchmarkKeccak256Write1MiB-4        300           4678746 ns/op         224.11 MB/s
BenchmarkKeccak256Sum-4          1000000              3421 ns/op        306437.18 MB/s
BenchmarkKeccak384Write1MiB-4        200           6685804 ns/op         156.84 MB/s
BenchmarkKeccak384Sum-4           500000              3363 ns/op        311790.40 MB/s
BenchmarkKeccak512Write1MiB-4        200           9324667 ns/op         112.45 MB/s
BenchmarkKeccak512Sum-4           500000              2703 ns/op        387898.02 MB/s
```

**Comment by fasterthanlime** [comment]
A few notes about these benchmarks:

  * There's 2 decoupled steps for every algo: Write and Sum, from the golang Hash interface. Since Keccak sponges have arbitrary output length, `Sum` is simulated by reading 16 bytes of output via `io.ReadAll` (which might be slower than real-world usage). The `ShakeSum` interface provided by `x/crypto/sha3` generates less garbage than the standard `Hash` interface (less GC pressure).
  * The md5 implementation that ships with Go is probably a highly-optimized implementation (probably part of openssl?), so the benchmark is a bit unfair
  * steakknife's keccak cgo package is based on the version of Keccak submitted to NIST while SHA-3 was still being picked. It's probable that the Keccak team has optimized their own implementation since.
  * In particular, my `BenchmarkKeccak128` bench case, that uses steakknife's package, uses the generic Keccak implementation, and not the optimized Shake128 implementation provided by the Keccak authors.
  * The `BenchmarkShake128` bench case uses `x/crypto/sha3`, which *sounds* like a pure Go implementation.

TL;DR need to build & run Keccak's optimized Shake128 routines to see if it's really going to be 2 to 3x slower than MD5 (they claim their implementation is faster than SHA1/MD5 now).

Also, it's worth noting that butler's operations are mostly IO-bound, limited by disk read speed, for *most* push operations (except the first one, in which case it's probably limited by upload speed), and a lot of things happen in parallel (different goroutines), so even if the hash was 3x slower, it doesn't mean butler will get 3x slower.
